### ISSUE TITLE
[Feature] Support composite learning rate scheduler

### ISSUE THREAD

**Issue by selmadeac** [open-mmlab/mmcv#2672]
### What is the feature?

Support composite learning rate scheduler

### Any other context?

Hi everyone!

I tried to create this composite lr scheduler. however it does not behave as expected at all. is here a way to use multiple LR's?
Also. I thought the get_lr is constantly called at each iteration. however the print is printed only once.
Can someone please give me some hints on what is actually going on?

@HOOKS.register_module()
class StepCycLrUpdaterHook(LrUpdaterHook):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.cos1 = StepLrUpdaterHook(step=[3,8],warmup='linear', warmup_iters=2000, by_epoch=False,   warmup_ratio=1.0 / 1000)

        self.cos3 = CyclicLrUpdaterHook(target_ratio=(1, 1e-4),warmup='linear', warmup_iters=2000,warmup_ratio=1.0 / 1000,cyclic_times=1,by_epoch=False,step_ratio_up=0.4)

    def get_lr(self, runner, base_lr):
        #print(runner.epoch,"runner.epoch")
        #print(runner.max_iters ,"self.max_iter_per_phase  ")
        if runner.epoch<8:
            return self.cos1.get_lr(runner,base_lr)
        else:
            return self.cos3.get_lr(runner,base_lr)

**Comment by selmadeac** [comment]
Solved: I think one cannot define more than one LrUpdaterHook for training.

I did rewrite the learning rates inside my class:
@HOOKS.register_module()
class CenterformerStepCycLrUpdaterHook(LrUpdaterHook):

    def __init__(self, **kwargs):  
        self.step=[4,8]
        self.gamma=0.1
        self.min_lr=1e-7

        self.target_ratio = (1, 1e-4)
        self.cyclic_times = 1
        self.step_ratio_up = 0.4
        self.gamma = 1
        self.max_iter_per_phase = None
        self.lr_phases: list = []  # init lr_phases
        # validate anneal_strategy

        anneal_strategy = 'cos'
        if anneal_strategy == 'cos':
            self.anneal_func: Callable[[float, float, float],
                                       float] = annealing_cos
        elif anneal_strategy == 'linear':
            self.anneal_func = annealing_linear

        

        self.split_epoch =  8

        

        super().__init__(**kwargs)
        assert not self.by_epoch, \
            'currently only support "by_epoch" = False'

    def get_lr(self, runner, base_lr):
        #print(runner.epoch,"runner.epoch")
        #print(runner.max_iters ,"self.max_iter_per_phase  ")
        if runner.epoch<self.split_epoch:
           
            progress = runner.epoch if self.by_epoch else runner.iter
            #print(progress,"progress")
            # calculate exponential term
            if isinstance(self.step, int):
                exp = progress // self.step
            else:
                exp = len(self.step)
                for i, s in enumerate(self.step):
                    if progress < s:
                        exp = i
                        break

            lr = base_lr * (self.gamma**exp)
            if self.min_lr is not None:
                # clip to a minimum value
                lr = max(lr, self.min_lr)
            return lr
        else:
            #print(runner.max_iters ,"self.max_iter_per_phase  ")
            curr_iter = runner.iter % self.max_iter_per_phase  # type: ignore
            curr_cycle = runner.iter // self.max_iter_per_phase  # type: ignore
            # Update weight decay
            scale = self.gamma**curr_cycle

            for (start_iter, end_iter, start_ratio, end_ratio) in self.lr_phases:
                if start_iter <= curr_iter < end_iter:
                    # Apply cycle scaling to gradually reduce the difference
                    # between max_lr and base lr. The target end_ratio can be
                    # expressed as:
                    # end_ratio = (base_lr + scale * (max_lr - base_lr)) / base_lr
                    # iteration: 0-iter_up_phase:
                    if start_iter == 0:
                        end_ratio = 1 - scale + end_ratio * scale
                    # iteration: iter_up_phase-self.max_iter_per_phase
                    else:
                        start_ratio = 1 - scale + start_ratio * scale
                    progress = curr_iter - start_iter
                    return self.anneal_func(base_lr * start_ratio,
                                            base_lr * end_ratio,
                                            progress / (end_iter - start_iter))

And now it works.
